packages/artifact/src/internal/download-http-client.ts

```diff
@@ -124,66 +194,117 @@ export class DownloadHttpClient {
       )
     }
 
-    if (isSuccessStatusCode(response.message.statusCode)) {
-      await this.pipeResponseToStream(
-        response,
-        stream,
-        isGzip(response.message.headers)
-      )
-    } else if (isRetryableStatusCode(response.message.statusCode)) {
-      warning(
-        `Received http ${response.message.statusCode} during file download, will retry ${artifactLocation} after 10 seconds`
-      )
-      // if an error is encountered, dispose of the http connection, and create a new one
+    const backOffUsingRetryValue = async (
+      retryAfterValue: number
+    ): Promise<void> => {
       this.downloadHttpManager.disposeAndReplaceClient(httpClientIndex)
-      await new Promise(resolve =>
-        setTimeout(resolve, getRetryWaitTimeInMilliseconds())
+      info(
+        `Backoff due to too many requests, retry #${retryCount}. Waiting for ${retryAfterValue} milliseconds before continuing the download`
       )
-      const retryResponse = await client.get(artifactLocation)
-      if (isSuccessStatusCode(retryResponse.message.statusCode)) {
-        await this.pipeResponseToStream(
+      await new Promise(resolve => setTimeout(resolve, retryAfterValue))
+      info(
+        `Finished backoff due to too many requests for retry #${retryCount}, continuing with download`
+      )
+      return
+    }
+
+    // keep trying to download a file until a retry limit has been reached
+    while (retryCount <= retryLimit) {
+      let response: IHttpClientResponse
+      try {
+        response = await makeDownloadRequest()
+      } catch (error) {
+        // if an error is catched, it is usually indicative of a timeout so retry the download
+        info('An error has been caught, while attempting to download a file')
+        // eslint-disable-next-line no-console
+        console.log(error)
+
+        // increment the retryCount and use exponential backoff to wait before making the next request
+        incrementAndCheckRetryLimit()
+        await backoffExponentially()
+        continue
+      }
+
+      if (isSuccessStatusCode(response.message.statusCode)) {
+        // The body contains the contents of the file however calling response.readBody() casues all the content to be converted to a string
+        // which can cause some gzip encoded data to lost
+        // Instead of using response.readBody(), response.message is a readablestream that can be directly used to get the raw body contents
+        await this.pipeResponseToFile(
           response,
-          stream,
+          destinationStream,
           isGzip(response.message.headers)
         )
+        return
+      } else if (isThrottledStatusCode(response.message.statusCode)) {
+        info(
+          'A 429 response code has been recieved when attempting to download an artifact'
```

Comment:
Nit, typo: should be `received`

Reply:
Quite a few spelling bloopers 😑 My spelling extension has been off for some reason